### PR TITLE
UCP/WIREUP/BW: mask RMA BW lanes by device

### PR DESCRIPTION
**Pull request by hoopoepg**
- BW lanes selection shifted into own routine
- RMA BW lane selection refactored to use new routine
  and unified with AM BW lane selection

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5951/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@brminich could you look at this?

File: src/ucp/wireup/select.c [hunk 1]
```diff
@@ -951,6 +927,64 @@ static ucs_status_t ucp_wireup_add_am_bw_lanes(ucp_ep_h ep, const ucp_ep_params_
...
+    ucp_wireup_select_bw_info_t bw_info;
+    ucp_lane_index_t lane_desc_idx;
+    ucp_rsc_index_t rsc_index;
+    unsigned addr_index;
+
+    /* Check if we need active messages, for wireup */
+    if (!(ucp_ep_get_context_features(ep) & UCP_FEATURE_TAG) ||
+        (ep_init_flags & UCP_EP_INIT_FLAG_MEM_TYPE)) {
+        return UCS_OK;
+    } else if (ep->worker->context->config.ext.max_eager_lanes < 2) {
```
Comment (brminich) [inline src/ucp/wireup/select.c]: Maybe better have one `if`?

Comment (hoopoepg) [reply src/ucp/wireup/select.c]: yep, why not...
done

**Comment by brminich** [comment]
:+1:

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5963/ for details (Mellanox internal link).

File: src/ucp/wireup/select.c [hunk 2]
```diff
@@ -856,85 +866,51 @@ static double ucp_wireup_am_bw_score_func(ucp_context_h context,
...
+    num_lanes          = 0;
+    md_map             = bw_info->md_map;
+    local_dev_bitmap   = bw_info->local_dev_bitmap;
+    remote_dev_bitmap  = bw_info->remote_dev_bitmap;
 
     /* lookup for requested number of lanes or limit of MD map
      * (we have to limit MD's number to avoid malloc in
      * memory registration) */
-    for (; (num_lanes < ep->worker->context->config.ext.max_eager_lanes) &&
+    for (; (num_lanes < bw_info->max_lanes) &&
```
Comment (yosefe) [inline src/ucp/wireup/select.c]: maybe it should be a while loop?

File: src/ucp/wireup/select.c [hunk 3]
```diff
@@ -856,85 +866,51 @@ static double ucp_wireup_am_bw_score_func(ucp_context_h context,
...
-                                             &criteria, -1, -1,
+                                             &bw_info->criteria, -1, -1,
                                              local_dev_bitmap, remote_dev_bitmap,
                                              0, &rsc_index, &addr_index, &score);
         if (status != UCS_OK) {
             break;
         }
 
-        is_proxy = ucp_wireup_is_lane_proxy(ep, rsc_index,
+        is_proxy = allow_proxy &&
```
Comment (yosefe) [inline src/ucp/wireup/select.c]: we need proxy only for AM lanes, not for RMA lanes

File: src/ucp/wireup/select.c (hunk 2 above)
Comment (hoopoepg) [reply src/ucp/wireup/select.c]: fixed

File: src/ucp/wireup/select.c (hunk 3 above)
Comment (hoopoepg) [reply src/ucp/wireup/select.c]: right, ```allow_proxy``` argument enables proxy for AM lanes only

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5969/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5987/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@yosefe squash changes?

**Comment by hoopoepg** [comment]
@yosefe good to go? or squash changes?

**Comment by yosefe** [comment]
@hoopoepg pls squash, also next time would be better to add more informative commit message instead of `..title.. (n)` so would not need to squash

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5998/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6000/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@yosefe squashed

**Pull request merged**